### ISSUE TITLE
Improving consistency for testing

### ISSUE THREAD

**Issue by U007D** [withoutboats/failure#14]
## TL;DR
can `Failure` introduce a consistent interface across all Fail types that implements `PartialEq` so errors can be compared simply and literally?  This will improve the ergonomics and productivity around unit testing, and possibly other scenarios as well.

## Motivation

As a BDD developer, the first thing I do when developing new code is to write a unit_test.  Rust's lack of consistent API around errors makes writing unit tests less ergonomic than perhaps they should be.  Here's a simple example of what I would like to write (which works, but only for std::io::Error):

```rust
#[test]
fn passing_bad_param_results_in_not_found_error() {
    assert_eq!(some_func(bad_param).err().kind(), std::io::ErrorKind::NotFound);
}
```

I've run into what I consider to be major ergonomic issues with `std::error::Error`'s and unit testing.  I haven't found a consistent way to work with errors that doesn't involve a lot of pattern-matching boilerplate.  (Adapted from a [posting](https://users.rust-lang.org/t/handling-rust-errors-concisely-consistently/13354) I made about a month ago:

## No consistent basis with which to work with Errors
(inconsistencies or issues in bold)

### using the Standard Library
* e.g. `std::io::Error`:
  - **is a struct encapsulating an enum**
  - **does not** implement PartialEq
  - **does** implement a kind() method
  - **does** implements ErrorKind variants which implement PartialEq

* e.g. `std::fmt::Error`:
  - **is a** (marker) **struct**
  - **does** implement PartialEq (among other traits)
  - **does not** implement a kind() method
  - **does not** define ErrorKind variants
  
* e.g. `std::io::CharsError`
  - **is an enum**
  - **does not** implement PartialEq
  - **does not** implement a kind() method
  - **does not** define ErrorKind variants

* `std::error::Error`
  - is a trait
  - **does not** mandate/require errors implement PartialEq
  - **does not** mandate/require a kind() method
  - **does hot** mandate/require ErrorKind variants
  
### using error-chain  
* `error-chain`
  - is Rust's defacto-standard error-handling crate
  - **does not** implement PartialEq
  - **does** implement a kind() method
  - **does** implement ErrorKind variants which **do not** implement PartialEq
  
### using pattern matching
  - pattern-matching (as opposed to via PartialEq) is often cited as the recommended way to distinguish Errors, given these numerous inconsistencies).  There are crates (such as [matches](https://crates.io/crates/matches)) for reducing boilerplate when pattern-matching
  - unfortunately, pattern matching `Error`s **adds boilerplate, obfuscates code and is an incomplete solution**. e.g.:
  ```
  fn string_validator_ref_handles_empty_input() {
    // GivenNotes:
    let input = String::new();
    let expected_result = ErrorKind::ValueNone;

    // When some operation which generates an error occurs
    let result = (&input).validate_ref::<NonEmptyStringValidator>().err().unwrap()
  ```

  ```
  /* What I would *like* to be able to write */

  // Then ensure the expected error was returned
    assert_eq!(*result, expected_result);
  }
  ```
  ```
  /* Reality */
  
  // Then ensure the expected error was returned
    assert!(matches!(*result, expected_result); //Error: pattern unreachable (match doesn't work this way; unintuitive at this level)
    assert!(matches!(*result, val if val == expected_result)); //not ergonomic; Error without PartialEq (e.g. error-chain): binary operation `==` cannot be applied to type `error::ErrorKind`
    assert!(matches!(*result, ErrorKind::ValueNone); //works, but without variable support, important scenarios become unrepresentable:
        // suppose two methods must yield same result.  To ensure consistency through refactorings, a test is developed:
        // assert_eq!(matches(FuncA(badInput), FuncB(badInput)) //not possible, as per above; only constant variants w/PartialEq implemented can be used ergonomically
  ```
When considering Error equality, I believe it is out of scope for  to consider the *semantics* of errors (e.g. "Are two instances of SomeError::OtherError equal?"  Assuming ::OtherError is a simple unit-struct variant, yes.  The fact that ::OtherError could be used by a library or application as a 'catch-all' to represent a variety of error conditions in the real world is out of scope for this proposal.  It is expected that the user of that error type will understand when a comparison is a meaningful operation). 

Given that the `Error` type is freqently implemented using different types (marker structs, encapsulating structs, enums, and others) the only dependencies we should take are on its API.  To that end, I would love to see a consistent `.kind()` method (or something else which serves the purpose of providing the error variant) which exists and is properly defined for every `Fail`ure.

**Comment by withoutboats** [comment]
I would like for error types to implement `PartialEq`, and for `Error` to, but there are problems. The core issue issue is that `PartialEq` is not object safe. This means we can't use it as a supertrait on `Fail`, and we can't really implement it for `Error`, which has a trait object inside it.

I have a vision of a solution, which requires specialization (an unstable feature). With specialization, we could support generating an (unsafe) function pointer for PartialEq *if* the type implements `PartialEq`, and not otherwise. Then, if the type ids match & there is a PartialEq implementation, we downcast both errors and dynamically call the PartialEq impl. (This conforms to PartialEq because PartialEq does not require reflexivity.)

In that sort of solution, we wouldn't *require* `PartialEq`, but Errors can be compared if their Fail type does implement it.

There's also nothing to stop users from deriving PartialEq for their Fail types now, whereas error-chain does not allow you to implement PartialEq for your error type at all.

**Comment by U007D** [comment]
Thanks for this, @withoutboats.  I have a couple of thoughts, but want to spend a few days to ramp up on object safety first to know how viable my ideas might be before posting.

**Comment by U007D** [comment]
I've been delving into the concept of Object Safety, its reason for being, history, etc.  I can't say I understand it all, and have many questions about why some limitations haven't been eliminated by compile-time trait object plumbing, but these are far outside the scope of this Issue.

So for this issue, even "dreaded" (from a matching consistency perspective) std::io::Error's discriminants ultimately boil down to either an i32 (`Os` variant) or a std::io::ErrorKind (`Simple` and `Custom` variants), both of which implement PartialEq.  Simpler error definitions such as `std::fmt::Error` and `std::str::Utf8Error`, are enums in their own right, often directly deriving PartialEq.

So my latest thinking is about whether Fail::Error could have a `.PartialEqErrorValueUsableForConsistentComparison()` method (the name is a work in progress ;)) which would return the error type's PartialEq discriminant (perhaps `std::mem::Discriminant`?  `std::mem::Discriminant` is actually generic, so more Object Safety concerns, but not sure how much of an issue that is for Fail::Error) so that the Error's user could then be compared consistently.

**Comment by shepmaster** [comment]
@U007D FWIW, `PartialEq` is the reason I use [quick-error](http://tailhook.github.io/quick-error/quick_error/index.html).

**Comment by U007D** [comment]
@shepmaster although better than the then-alternatives, I found quick-error still too verbose...
